rust/src/dataset/transaction.rs

```diff
@@ -0,0 +1,555 @@
+//! manifest.
+
+use std::{collections::HashSet, sync::Arc};
+
+use crate::{
+    datatypes::Schema,
+    format::{pb, Fragment, Manifest},
+};
+
+use super::{feature_flags::apply_feature_flags, ManifestWriteConfig};
+use crate::{Error, Result};
+
+/// A change to a dataset that can be retried
+///
+/// This contains enough information to be able to build the next manifest,
+/// given the current manifest.
+#[derive(Debug, Clone)]
+pub struct Transaction {
+    /// The version of the table this transaction is based off of. If this is
+    /// the first transaction, this should be 0.
+    pub read_version: u64,
+    pub uuid: String,
+    pub operation: Operation,
+    pub tag: Option<String>,
+}
+
+/// An operation on a dataset.
+#[derive(Debug, Clone)]
+pub enum Operation {
+    /// Adding new fragments to the dataset. The fragments contained within
+    /// haven't yet been assigned a final ID.
+    Append { fragments: Vec<Fragment> },
+    /// Updated fragments contain those that have been modified with new deletion
+    /// files. The deleted fragment IDs are those that should be removed from
+    /// the manifest.
+    Delete {
+        updated_fragments: Vec<Fragment>,
+        deleted_fragment_ids: Vec<u64>,
+        predicate: String,
+    },
+    /// Overwrite the entire dataset with the given fragments. This is also
+    /// used when initially creating a table.
+    Overwrite {
+        fragments: Vec<Fragment>,
+        schema: Schema,
+    },
+    /// A new index has been created.
+    CreateIndex,
+    /// Data is rewritten but *not* modified. This is used for things like
+    /// compaction or re-ordering. Contains the old fragments and the new
+    /// ones that have been replaced.
+    Rewrite {
+        old_fragments: Vec<Fragment>,
+        new_fragments: Vec<Fragment>,
+    },
+    /// Merge a new column in
+    Merge {
+        fragments: Vec<Fragment>,
+        schema: Schema,
+    },
+}
+
+impl Transaction {
+    pub fn new(read_version: u64, operation: Operation, tag: Option<String>) -> Self {
+        let uuid = uuid::Uuid::new_v4().hyphenated().to_string();
+        Self {
+            read_version,
+            uuid,
+            operation,
+            tag,
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        match &self.operation {
+            Operation::Append { .. } => "Append",
+            Operation::Delete { .. } => "Delete",
+            Operation::Overwrite { .. } => "Overwrite",
+            Operation::CreateIndex => "CreateIndex",
+            Operation::Rewrite { .. } => "Rewrite",
+            Operation::Merge { .. } => "Merge",
+        }
+    }
+
+    /// Returns true if the transaction cannot be committed if the other
+    /// transaction is committed first.
+    pub fn conflicts_with(&self, other: &Self) -> bool {
+        // TODO: this assume IsolationLevel is Serializable, but we could also
+        // support Snapshot Isolation, which is more permissive.
+        match &self.operation {
```

Comment:
uber nit (ticket for now is fine): I wonder if there is a way to test this operation matrix is diagonally symmetrical. Just so that when we add new ones we'd know if we mess up

Reply:
Tbh initially I didn’t think the matrix would be symmetric. But perhaps it is. I’ll think about this and verify in the morning

Reply:
Right now the only one that's not symmetric is Append-Delete. We allow Append after Delete, but not Delete after Append, since the delete might have wanted to remove data the append just added. There are looser isolation levels that would make this symmetric. Databricks calls this `WriteSerializable`, https://docs.databricks.com/en/optimizations/isolation-level.html. Iceberg calls this `Snapshot Isolation` https://iceberg.apache.org/javadoc/0.11.0/org/apache/iceberg/IsolationLevel.html

Reply:
I see makes sense.